### PR TITLE
SONARJAVA-6490 Implement S8910: Interfaces with @Mapper should have @DaoFactory methods

### PR DESCRIPTION
**Pull request by romainbrenguier**
## Summary
Implements rule S8910 to detect interfaces annotated with `@Mapper` that don't contain at least one `@DaoFactory` method. https://github.com/SonarSource/rspec/pull/7151

## Implementation
- Rule type: CODE_SMELL
- Severity: Major
- Detection: Checks interfaces with `@Mapper` annotation for presence of `@DaoFactory` methods
- Follows patterns from similar annotation-checking rules (e.g., S7180)

## Status
WIP - Implementation incomplete after initial attempts. Requires review and completion.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

----
## Summary by Gitar

- **Rule Metadata:**
  - Added rule S8910 to `Sonar_way_profile.json` and `Sonar_agentic_AI_profile.json`.
  - Created `S8910.html` and `S8910.json` for rule documentation and configuration.
- **CI/Test Infrastructure:**
  - Created `diff_S8910.json` for `autoscan` verification.
  - Updated `JavaAgenticWayProfileTest` to validate the new rule integration.
  - Updated `build.yml` to use Java 26.
- **Implementation Assets:**
  - Added stub annotations `Mapper.java` and `DaoFactory.java` for test source compilation.
  - Provided `MapperWithoutDaoFactoryCheckSample.java` and `non-compiling/checks/MapperWithoutDaoFactoryCheckSample.java` covering edge cases.
  - Implemented `MapperWithoutDaoFactoryCheck.java` and corresponding unit tests.

<sub>This will update automatically on new commits.</sub>

### CODE REVIEW

File: fix-unit-tests/test-output.log
Comment (romainbrenguier) [inline fix-unit-tests/test-output.log:1]: Delete this file. It should not be comited in this PR.

File: java-checks-test-sources/default/src/main/java/checks/MapperWithoutDaoFactoryCheckSample.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  interface BaseInterface {
+    @DaoFactory
+    BaseDao baseDao();
+  }
+
+  interface BaseDao {
+  }
+
+  interface RegularInterface {
+  }
```
Comment (NoemieBenard) [inline java-checks-test-sources/default/src/main/java/checks/MapperWithoutDaoFactoryCheckSample.java]: I think this is unnecessary as it is never used and is already tested by the previous sample.
```suggestion
```

File: java-checks-test-sources/default/src/main/java/checks/MapperWithoutDaoFactoryCheckSample.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
+package checks;
+
+import com.datastax.oss.quarkus.runtime.api.mapper.Mapper;
+import com.datastax.oss.quarkus.runtime.api.mapper.DaoFactory;
+
+class MapperWithoutDaoFactoryCheckSample {
+
+  @Mapper
+  interface EmptyMapper { // Noncompliant {{Add at least one "@DaoFactory" method to this "@Mapper" interface.}}
+  }
```
Comment (NoemieBenard) [inline java-checks-test-sources/default/src/main/java/checks/MapperWithoutDaoFactoryCheckSample.java]: minor: I would add at least one precise location test

**Review by NoemieBenard — APPROVED** [review]
Implementation looks good to me, just some comments about the samples

File: java-checks/src/main/java/org/sonar/java/checks/MapperWithoutDaoFactoryCheck.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  private static boolean hasDaoFactoryMethod(Symbol.TypeSymbol typeSymbol, Set<Symbol.TypeSymbol> visited, boolean hasExplicitSuperInterfaces) {
+    if (!visited.add(typeSymbol)) {
+      return false;
+    }
+
+    if (typeSymbol.memberSymbols().stream()
+      .filter(Symbol::isMethodSymbol)
+      .map(Symbol.MethodSymbol.class::cast)
+      .anyMatch(MapperWithoutDaoFactoryCheck::hasDaoFactoryAnnotation)) {
```
[the remark this answers is not in the document]
Comment (romainbrenguier) [reply java-checks/src/main/java/org/sonar/java/checks/MapperWithoutDaoFactoryCheck.java:61]: Good catch! Simplified the implementation to remove the redundant recursion, visited set, and the hasExplicitSuperInterfaces parameter passing. Since `superTypes()` already returns the full transitive closure (via `JUtils.collectSuperTypes`), we can check each supertype's members directly in a single pass. The FP-avoidance for unknown supertypes now only evaluates at the top level where `hasExplicitSuperInterfaces` is properly set from `classTree.superInterfaces()`, which is exactly what's needed.
